odoo/models.py

```diff
@@ -6017,6 +6017,7 @@ def diff(self, other):
                         result[name] = field.convert_to_onchange(self[name], record, {})
                     else:
                         # x2many fields: serialize value as commands
+                        record[name].with_prefetch()
```

Comment:
Dude, this is black magic to *me*!
* The left part simply reads the value of field `name` on `record`.  It has no impact on the `line` records that appear in the snapshots in `self[name]`, as those have their own *prefetch set*, which is `record[name]` at the time the snapshot was made.
* The rest is even weirder.  `self.with_prefetch()` essentially returns `self` with an new *prefetch set*.  For this to have some side effect, you *must* access some data from the result.  But the result is not used here!

Did I miss something?
How did you come up with that?

Reply:
Originally I had
```python
                        if self[name]:
                            to_read = record.env[self[name][0]['<record>']._name]
                            for line_snapshot in self[name]:
                                to_read += line_snapshot['<record>']._origin
                            to_read.read()
```
Which works too, but it still makes more queries (for the same test I had 101 queries instead of 74 for the onchange on `account.move` (1409 before any fix))

Then I remembered the `with_prefetch` method exists; not really knowing what it is for
```python
    def with_prefetch(self, prefetch_ids=None):
        """ with_prefetch([prefetch_ids]) -> records

        Return a new version of this recordset that uses the given prefetch ids,
        or ``self``'s ids if not given.
        """
```
I read this as "if the ids are not given, prefetch the values of self.ids and put it in the cache". By doing so, it is put in the cache, and the later call to `Snapshot(line, subnames)` does not trigger a read.

Reply:
@rco-odoo I put some profiles made with `from odoo.tools.misc import profile` in the Discord discussion if needed.

Reply:
> Then I remembered the `with_prefetch` method exists; not really knowing what it is for
> ```python
>     def with_prefetch(self, prefetch_ids=None):
>         """ with_prefetch([prefetch_ids]) -> records
> 
>         Return a new version of this recordset that uses the given prefetch ids,
>         or ``self``'s ids if not given.
>         """
> ```
> I read this as "if the ids are not given, prefetch the values of self.ids and put it in the cache". (...)

This is not at all what the docstring says.  It says that the result is a recordset like ``self``, which is assigned the given prefetch ids, i.e., the candidates to consider *when you have to prefetch*.

You should have read the implementation of the method, which is only 3 lines of code, and clearly just browses a recordset.  This does nothing more than creating a Python object.  There is nothing that calls `read` or anything similar.

Reply:
Right. _Only 3 lines of code._
```python
    def with_prefetch(self, prefetch_ids=None):
        """ with_prefetch([prefetch_ids]) -> records

        Return a new version of this recordset that uses the given prefetch ids,
        or ``self``'s ids if not given.
        """
        if prefetch_ids is None:
            prefetch_ids = self._ids
        return self._browse(self.env, self._ids, prefetch_ids)
```
Then 
```python
    @classmethod
    def _browse(cls, env, ids, prefetch_ids):
        """ Create a recordset instance.

        :param env: an environment
        :param ids: a tuple of record ids
        :param prefetch_ids: a collection of record ids (for prefetching)
        """
        records = object.__new__(cls)
        records.env = env
        records._ids = ids
        records._prefetch_ids = prefetch_ids
        return records
```

`_prefetch_ids` seems to be used only in `_in_cache_without()`, `_origin()`, `__iter__()` and `with_env()`.
None of these do something explicit with it.
```python
    def with_env(self, env):
        return self._browse(env, self._ids, self._prefetch_ids)
```
```python
    @property
    def _origin(self):
        ids = tuple(origin_ids(self._ids))
        prefetch_ids = IterableGenerator(origin_ids, self._prefetch_ids)
        return self._browse(self.env, ids, prefetch_ids)
```
```python
    def __iter__(self):
        for id in self._ids:
            yield self._browse(self.env, (id,), self._prefetch_ids)
```
```python
    @api.model
    def _in_cache_without(self, field, limit=PREFETCH_MAX):
        kind = bool(self.id)
        recs = self.browse(unique(self._prefetch_ids))
        ids = [self.id]
        for record_id in self.env.cache.get_missing_ids(recs - self, field):
            if bool(record_id) != kind:
                continue
            ids.append(record_id)
            if limit and limit <= len(ids):
                break
        return self.browse(ids)
```

It is also used in 
```python
def prefetch_many2one_ids(record, field):
    records = record.browse(record._prefetch_ids)
    ids = record.env.cache.get_values(records, field)
    return unique(id_ for id_ in ids if id_ is not None)
```
```python
def prefetch_x2many_ids(record, field):
    records = record.browse(record._prefetch_ids)
    ids_list = record.env.cache.get_values(records, field)
    return unique(id_ for ids in ids_list for id_ in ids)
```

Which are used in `convert_to_record()`

Sorry If I don't find it the most straightforward 3 lines ever.

__________

Now this said, the value set for `_prefetch_ids` is used here:
```
  File "/home/odoo/git/odoo/odoo/odoo/models.py", line 6140, in onchange
    result['value'] = snapshot1.diff(snapshot0)
  File "/home/odoo/git/odoo/odoo/odoo/models.py", line 6035, in diff
    line_diff = line_snapshot.diff(Snapshot(line, subnames))
  File "/home/odoo/git/odoo/odoo/odoo/models.py", line 5978, in __init__
    self.fetch(name)
  File "/home/odoo/git/odoo/odoo/odoo/models.py", line 5988, in fetch
    self[name] = record[name]
  File "/home/odoo/git/odoo/odoo/odoo/models.py", line 5643, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/home/odoo/git/odoo/odoo/odoo/fields.py", line 1009, in __get__
    recs._fetch_field(self)
  File "/home/odoo/git/odoo/odoo/odoo/models.py", line 2947, in _fetch_field
    self._read(fnames)
```
I don't know why
```python
                        record[name].with_prefetch()
                        print(list(record[name]._prefetch_ids))
```
and 
```python
                        record[name] = record[name].with_prefetch()
                        print(list(record[name]._prefetch_ids))
```
yield the same result, but both have `_prefetch_ids` set.

______

I have another funny one for you now.
This has the same effect.
```suggestion
                        record[name]._prefetch_ids
```